### PR TITLE
IPC4 error handling

### PR DESCRIPTION
**Pull request by lyakh**
switch IPC4 internal error codes to POSIX, also fix several apparent bugs

### CODE REVIEW

File: src/include/sof/ipc/common.h [hunk 1]
```diff
@@ -90,6 +90,13 @@ struct ipc {
 #define ipc_get_drvdata(ipc) \
 	((ipc)->private)
 
+struct ipc_core_ctx {
+	int error;
```
Comment (lgirdwood) [inline src/include/sof/ipc/common.h:94]: this may not always be an error, but its more of a reply with context. maybe call it ipc4_reply.

File: src/arch/xtensa/ipc.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#include <sof/ipc/common.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cpu.h>
+#include <xtos-structs.h>
+
+struct ipc_core_ctx **arch_ipc_get(void)
+{
+	struct core_context *ctx = (struct core_context *)cpu_read_threadptr();
+
+	return &ctx->ipc;
```
Comment (lgirdwood) [inline src/arch/xtensa/ipc.c:21]: Why do we need for xtos ?

**Review by lgirdwood — COMMENTED** [review]
Client API looks simple and will allow compatibility. Have you tried with Zephyr ?
@mwasko @marcinszkudlinski any comments ?

File: src/ipc/ipc4/helper.c [hunk 3]
```diff
@@ -751,29 +886,3 @@ struct comp_dev *ipc4_get_comp_dev(uint32_t comp_id)
 
 	return icd->cd;
 }
-
-int ipc4_add_comp_dev(struct comp_dev *dev)
```
Comment (ranj063) [inline src/ipc/ipc4/helper.c:755]: why is this removed? doesnt seem related to the change in this patch or I am not able to connect immediately

File: src/arch/xtensa/ipc.c (hunk 2 above)
Comment (lyakh) [reply src/arch/xtensa/ipc.c:21]: It is currently possible to build IPC4 with XTOS, right? And I think that configuration is or was also used for the initial development / testing? So I thought I shouldn't break that functionality

File: src/ipc/ipc4/helper.c (hunk 3 above)
Comment (lyakh) [reply src/ipc/ipc4/helper.c:755]: That function isn't called anywhere outside the file where it's defined, so I made it static

File: src/arch/xtensa/ipc.c (hunk 2 above)
Comment (lgirdwood) [reply src/arch/xtensa/ipc.c:21]: Lets use Zephyr for this work

File: src/include/sof/ipc/common.h (hunk 1 above)
Comment (mwasko) [inline src/include/sof/ipc/common.h:94]: How in this design you are going to guarantee that ipc error code will not be overwritten prior to returning it in the reply message?

File: src/arch/xtensa/ipc.c (hunk 2 above)
Comment (lyakh) [reply src/arch/xtensa/ipc.c:21]: @lgirdwood well, sure, I wouldn't mind that, but what exactly does that mean? Does that mean that it should only be possible to build SOF with IPC4 with Zephyr? So we should disable building IPC4 with XTOS? If we don't do that and keep the possibility to build SOF with IPC4, then how would error handling work there?

File: src/include/sof/ipc/common.h (hunk 1 above)
Comment (lyakh) [reply src/include/sof/ipc/common.h:94]: @mwasko overwritten by whom? That variable is only set when processing IPC messages and a new IPC request cannot arrive before the previous one has been completed and replied to. This is still the case with IPC4, right? And asynchronous IPC messages, sent from the firmware to the host don't use that variable

File: src/arch/xtensa/ipc.c (hunk 2 above)
Comment (lgirdwood) [reply src/arch/xtensa/ipc.c:21]: @lyakh IPC4 with XTOS is not needed.

**Comment by mwasko** [comment]
@lyakh can you explain why we need to change internal IPC4 error codes into POSIX? As far as I can tell the affected functions are only called on IPC/IDC request so it makes sense to me to return IPC specific status. The proposed change also increase memory consumption and complexity.

**Comment by lgirdwood** [comment]
@mwasko sorry, this is not about changing IPC4 codes to posix, but about preserving the IPC4 reply within a POSIX environment so it can be sent correctly to the host drivers for every IPC. The PR naming probably does not help.

**Comment by mwasko** [comment]
> @mwasko sorry, this is not about changing IPC4 codes to posix, but about preserving the IPC4 reply within a POSIX environment so it can be sent correctly to the host drivers for every IPC. The PR naming probably does not help.

I understand the idea, but this PR is a great example how cumbersome can be aligning with POSIX environment.

Since we are moving to Zephyr and we are no longer going to maintain generic kernel code (where POSIX makes sense), but instead of that SOF is going to be Audio application layer, then maybe we should consider to switch to application specific error codes within SOF audio domain.

**Comment by lyakh** [comment]
@mwasko I agree in principle, that we don't have to use POSIX error codes or any other APIs. But what I do think we have to do is use consistent APIs throughout our code. Yes, we have to interface to other systems: to Zephyr, to the Linux kernel, etc. There we have to define interfaces well and - if needed - convert between them. That's what we do e.g. with IPC3 triggers - we have IPC3 trigger commands, that we exchange with the user, in this case with the Linux kernel. And internally SOF defines its own trigger codes, and it cleanly converts between the two sets at the interface.
The same has to be done with return codes. Yes, IPC4 defines its set of error codes and SOF has to use them to communicate with IPC4 counterparts. That's it. We don't need to modify parts of SOF to use those codes for any other purposes e.g. for returning them from functions. The whole of SOF uses POSIX error codes so that's what we should continue to use consistently. Anything else we only use to interface with respective other parties.
And in fact you can already see multiple examples of confusion, fixed by this PR. It's easy enough to find multiple cases where error codes are currently confused. Error codes from generic SOF functions aren't evaluated and converted to IPC4 codes but are instead propagated as is. Just for a single example look at the current version of `set_pipeline_state()` in ipc4/handler.c. In most places it returns IPC4 error codes but in one place it also returns the return code from `ipc_process_on_core()` directly, which is a generic SOF function and returns POSIX error codes, 0 or 1. That's just one example of the confusion that such API mixing is causing.

**Comment by mwasko** [comment]
> @mwasko I agree in principle, that we don't have to use POSIX error codes or any other APIs. But what I do think we have to do is use consistent APIs throughout our code. Yes, we have to interface to other systems: to Zephyr, to the Linux kernel, etc. There we have to define interfaces well and - if needed - convert between them. That's what we do e.g. with IPC3 triggers - we have IPC3 trigger commands, that we exchange with the user, in this case with the Linux kernel. And internally SOF defines its own trigger codes, and it cleanly converts between the two sets at the interface. 
The same has to be done with return codes. Yes, IPC4 defines its set of error codes and SOF has to use them to communicate with IPC4 counterparts. That's it. We don't need to modify parts of SOF to use those codes for any other purposes e.g. for returning them from functions. The whole of SOF uses POSIX error codes so that's what we should continue to use consistently.

I am okay with the error status conversion at IPC process level but to do that we need something better then POSIX error codes. The only reason you had to add context for ipc error is because POSIX is too generic and we are unable to translate it into IPC status code. Additionally the specific application error codes can significantly speed up the debug and issue root causing. 

> Anything else we only use to interface with respective other parties. And in fact you can already see multiple examples of confusion, fixed by this PR. It's easy enough to find multiple cases where error codes are currently confused. Error codes from generic SOF functions aren't evaluated and converted to IPC4 codes but are instead propagated as is. Just for a single example look at the current version of `set_pipeline_state()` in ipc4/handler.c. In most places it returns IPC4 error codes but in one place it also returns the return code from `ipc_process_on_core()` directly, which is a generic SOF function and returns POSIX error codes, 0 or 1. That's just one example of the confusion that such API mixing is causing.

I agree that IPC4 specific status as a return of generic functions like set_pipeline_state is not a way to go, but I would expect a meaningful set_pipeline_state error code that can be consumed both by IPC3 and IPC4 handlers (or any other caller).

My recommendation is to define of SOF application error code enum and apply it incrementally with new code implementation/refactors, starting with IPC handlers.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@lyakh I think we can probably close now ? Can always reopen if things get difficult.

**Comment by lyakh** [comment]
> @lyakh I think we can probably close now ? Can always reopen if things get difficult.

@lgirdwood This PR is fixing real bugs and as a result of the discussion herein it was agreed to try to fix those bugs differently. However, those bugs still haven't been fixed. We can close this PR and I can add a high priority bug report instead if you prefer.

File: src/arch/xtensa/ipc.c (hunk 2 above)
Comment (lyakh) [reply src/arch/xtensa/ipc.c:21]: @lgirdwood I just learned from @ujfalusi that work is ongoing for IPC4 with XTOS

Comment (ujfalusi) [reply src/arch/xtensa/ipc.c:21]: @lyakh, to clarify: testing is done with both xtos and zephyr, but I personally build SOF with XTOS for my local testing. The parts that I'm working with are way above anything xtos/zephyr and makes no difference for the end result.

Comment (lyakh) [reply src/arch/xtensa/ipc.c:21]: @ujfalusi ok, got it, but IIUC it might happen that IPC4 will not even build with XTOS beginning from some point. In fact maybe we should add a Kconfig dependency for that.

**Comment by lgirdwood** [comment]
Lets close now - we can always reopen if circumstances change.
